storage/indexer/src/db_indexer.rs

```diff
@@ -0,0 +1,396 @@
+    }
+
+    pub fn run(&self) {
+        loop {
+            let batch_opt = self
+                .receiver
+                .recv()
+                .expect("Failed to receive batch from DB Indexer");
+            if let Some(batch) = batch_opt {
+                self.db
+                    .write_schemas(batch)
+                    .expect("Failed to write batch to indexer db");
+            } else {
+                break;
+            }
+        }
+    }
+}
+
+pub struct DBIndexer {
+    pub db: Arc<DB>,
+    pub main_db_reader: Arc<dyn DbReader>,
+    config: InternalIndexerDBConfig,
+    sender: Sender<Option<SchemaBatch>>,
+    committer_handle: Option<thread::JoinHandle<()>>,
+}
+
+impl Drop for DBIndexer {
+    fn drop(&mut self) {
+        if let Some(handle) = self.committer_handle.take() {
+            self.sender
+                .send(None)
+                .expect("Failed to send None to DBIndexer committer");
+            handle
+                .join()
+                .expect("DBIndexer committer thread fails to join");
+        }
+    }
+}
+
+impl DBIndexer {
+    pub fn new(
+        db: Arc<DB>,
+        db_reader: Arc<dyn DbReader>,
+        config: &InternalIndexerDBConfig,
+    ) -> Self {
+        let (sender, reciver) = mpsc::channel();
+
+        let db_clone = db.clone();
+        let committer_handle = thread::spawn(move || {
+            let committer = DBCommitter::new(db, reciver);
+            committer.run();
+        });
+
+        Self {
+            db: db_clone,
+            main_db_reader: db_reader,
+            config: *config,
+            sender,
+            committer_handle: Some(committer_handle),
+        }
+    }
+
+    pub fn get_persisted_version(&self) -> Version {
+        // read the latest key from the db
+        let mut iter_res = self
+            .db
+            .iter::<InternalIndexerMetadataSchema>()
+            .expect("Cannot create db metadata iterator");
+        iter_res.seek_to_first();
+        iter_res
+            .next()
+            .map(|res| res.map_or(0, |(_, meta_value)| meta_value.expect_version()))
+            .unwrap_or_default()
+    }
+
+    pub fn event_enabled(&self) -> bool {
+        self.config.enable_event
+    }
+
+    pub fn transaction_enabled(&self) -> bool {
+        self.config.enable_transaction
+    }
+
+    fn get_internal_db_iter(
+        &self,
+        start_version: Version,
+        num_transactions: u64,
+    ) -> Result<Box<dyn Iterator<Item = Result<(Transaction, Vec<ContractEvent>)>> + '_>> {
+        let txn_iter = self
+            .main_db_reader
+            .get_transaction_iterator(start_version, num_transactions)?;
+        let mut event_vec_iter = self
+            .main_db_reader
+            .get_events_iterator(start_version, num_transactions)?;
+        let zipped = txn_iter.enumerate().map(move |(_idx, txn_res)| {
+            let txn = txn_res?;
+            let event_vec = event_vec_iter.next().unwrap_or(Ok(Vec::new()))?;
```

Comment:
no? this will make the indexer miss events?  should terminate when either indexer terminates?

Reply:
I only process the transaction before the synced_version. I should be able to assume the event always exists in this case?